Feature/add build test

JulieSchramm opened the pull request:
## DESCRIPTION OF CHANGES: 
Add ``build.sh`` to test directory to provide the build step for automated testing.  A ``README.md`` file is included with information on how to run the build.sh script.

## TESTS CONDUCTED: 
build.sh script tested on Cheyenne (Intel and GNU), Hera (Intel) and Jet (Intel).  All builds succeeded.

## DEPENDENCIES:
None

## DOCUMENTATION:
Provided in the README.md file.  Will be included in the User's Guide when testing is added.

Commit:
Add test/build.sh to build all executables for supported compilers on platforms Cheyenne, Hera and Jet.
Add test/README.md file for usage.
All builds succeed on Cheyenne.

Commit:
Merge branch 'develop' into feature/add_build_test

Commit:
Merge branch 'develop' into feature/add_build_test

gsketefian commented on test/build.sh:
```diff
@@ -0,0 +1,146 @@
+#!/bin/bash -l
```
Does it work without -l?  That would be better.

danrosen25 replied:
If possible I think it's best to design the script to be independent of the shell invocation selected. So it's best to eliminate any dependence on the environment settings at the time the script is called.

gsketefian commented:
How is build.sh different from the devbuild.sh that @mkavulich put together?

christinaholtNOAA commented on test/build.sh:
```diff
@@ -0,0 +1,146 @@
+if [ "$1" = "-h" ] ; then usage ; fi
+
+export machine=${1}
+machine=`echo "${machine}" | tr '[A-Z]' '[a-z]'`  # scripts in sorc need lower case machine name
```
In general, the `$()` syntax is easier to read/debug.

JulieSchramm replied:
Agree, will change.

christinaholtNOAA commented on test/build.sh:
```diff
@@ -0,0 +1,146 @@
+#-----------------------------------------------------------------------
+# Set some directories
+#-----------------------------------------------------------------------
+PID=$$
```
Just for my enrichment, what does this syntax do?

JulieSchramm replied:
``$$`` is a process ID (PID) used to create a unique output filename each time the script is run so the previous output is saved.

christinaholtNOAA replied:
Thanks, @JulieSchramm!

christinaholtNOAA commented on test/build.sh:
```diff
@@ -0,0 +1,146 @@
+    BUILD_DIR=${TOP_DIR}/build_${compiler}
+    BUILD_OUTPUT=${BUILD_DIR}/build.out
+    BIN_DIR=${TOP_DIR}/bin_${compiler}
+    EXEC_DIR=${BIN_DIR}/bin
```
Just curious why a subdirectory is needed.

JulieSchramm replied:
In the top-level CMakeLists.txt file, the install directory is set as ``bin`` under what  ``-DCMAKE_INSTALL_PREFIX`` is set to.  To make multiple builds and save all the executables for testing, I set ``-DCMAKE_INSTALL_PREFIX=../bin_${compiler}``, which creates ``../bin_intel/bin``, ``../bin_gnu/bin``, etc.

JulieSchramm commented:
The -l is handy if you call this script from crontab.  I can remove it, and
it can be added in the crontab file if someone wants to use it that way.

And how is the build.sh different from the devbuild.sh that @danrosen25 is
putting together?

On Mon, Jul 19, 2021 at 3:27 PM gsketefian ***@***.***> wrote:

> How is build.sh different from the devbuild.sh that @mkavulich
> <https://github.com/mkavulich> put together?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ufs-community/ufs-srweather-app/pull/156#issuecomment-882873056>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3WNU7OFWR4S45QPGR75FDTYSKFJANCNFSM5AUJXGPA>
> .
>

christinaholtNOAA commented on test/build.sh:
```diff
@@ -0,0 +1,146 @@
+      if [ -d "${BIN_DIR}" ] ; then rm -rf ${BIN_DIR} ; fi
+      mkdir ${BUILD_DIR}
+      cd ${BUILD_DIR}
+      cmake .. -DCMAKE_INSTALL_PREFIX=../bin_${compiler}
```
Can this line use `${BIN_DIR}` to reduce duplication of information?

JulieSchramm replied:
Yes!

christinaholtNOAA reviewed (approved):
Just a couple of very minor suggestions, and a quick question for my curiosity (hard to google).

JeffBeck-NOAA commented:
@JulieSchramm, @danrosen25's PR only improves the devbuild.sh script that @mkavulich created.  But I think @gsketefian is asking why we need this script in addition to devbuild.sh.

JulieSchramm commented:
The build.sh script is for testing.  It builds all executables for all supported compilers on a given machine.  It checks for the existence of all executables and writes a pass/fail message at the end.  This will be the build step for Christina et al.'s testing.

Commit:
Address Gerard's and Christina's comments:

- Remove -l from top of script
- Use ${BIN_DIR} for cmake command
- Use $() for easier debugging

gsketefian commented:
I'm wondering if it's possible to call the `devbuild.sh` script within the `build.sh` script introduced in this PR so that the same build script that is normally used to build the code is used to test the build, i.e. `build.sh` would be a wrapper around `devbuild.sh`.  @danrosen25 is adding a bunch of new arguments to `devbuild.sh` in PR #155.  If those changes aren't enough, `devbuild.sh` can probably be tweaked further so that it can be called by `build.sh` (instead of `build.sh` issuing its own `cmake` and `make` commands).

JulieSchramm commented:
I would leave the decision about build.sh vs devbuild.sh to @christinaholtNOAA, as the group at GSL will be setting up the automated testing.

christinaholtNOAA commented:
I have no strong opinions about one way or the other. If any users rely on `devbuild.sh` for building, I could see how calling it (instead of cmake directly) would essentially be testing an additional piece of used code, which is always better than not testing code that is used. Maybe that is enough to tip the scale toward calling `devbuild.sh` instead of using cmake?

danrosen25 commented:
Just a heads up the devbuild.sh changes will take arguments for
- `--platform=<PLATFORM>` , if omitted this will use the machine_setup.sh script
- `--compiler=<COMPILER>`, defaults to intel on cheyenne, hera, jet
- `--build-dir=<BUILD_DIR>`, the CMake build directory
- `--install-dir=<INSTALL_DIR>` the CMAKE_INSTALL_PREFIX directory, which is put into the build directory if you don't give a full path.

More can be added as needed. I'm trying to make a user|developer|sysadmin friendly build script. We needed a way to build different configurations of the model i.e. `--app=<APPLICATION>`

Commit:
Merge branch 'develop' into feature/add_build_test

Commit:
Use the devbuild.sh script for the build tests.

JulieSchramm commented:
All builds passed on Cheyenne, Jet and Hera.

JeffBeck-NOAA commented on test/build.sh:
```diff
@@ -0,0 +1,137 @@
+ENV_DIR=${TOP_DIR}/env
+#-----------------------------------------------------------------------
+# Arrray of all executables built
+#-----------------------------------------------------------------------
```
Typo: Arrray to Array.

JulieSchramm replied:
Good catch

JeffBeck-NOAA commented on test/build.sh:
```diff
@@ -0,0 +1,137 @@
+  done
+  #-----------------------------------------------------------------------
+  # Rename build and bin dirs for next build
+  #-----------------------------------------------------------------------
```
I'm confused by this comment block above.  What code does it apply to?

JulieSchramm replied:
Removed

Commit:
Merge branch 'develop' into feature/add_build_test

Commit:
Fix typo
Remove irrelevent comment block
